### PR TITLE
Add ReduceLROnPlateauScheduler feature

### LINKED ISSUE #1754: Implementation of ReduceLROnPlateau scheduler
## 🚀 Feature

While there might be some workarounds in using `ReduceLROnPlateau` with ignite's optimizer , it would be nicer to have a standalone feature like other schedulers in contrib.

[Relevant issue and code sample](https://github.com/pytorch/ignite/issues/462#issuecomment-771907302).

### PR DESCRIPTION
**Pull request by sadra-barikbin**
Fixes #1754 

Description:
Add ReduceLROnPlateauScheduler which basically is a wrapper around `torch.optim.lr_scheduler.ReduceLROnPlateau`. It supports `param_group_index` and `save_history` params.
Check list:

- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [x] Documentation is updated (if required)

### CODE REVIEW

**Comment by vfdev-5** [comment]
@sadra-barikbin thanks for the PR, can you run the following to fix code format: 
- `bash ./tests/run_code_style.sh install`
- `bash ./tests/run_code_style.sh fmt`

File: ignite/handlers/param_scheduler.py [hunk 1]
```diff
@@ -1393,6 +1393,137 @@ def simulate_values(cls, num_events: int, schedulers: List[_LRScheduler], **kwar
...
+    def _reduce_lr(self, epoch):
+        for i, param_group in enumerate(self.optimizer_param_groups):
+            old_lr = float(param_group["lr"])
+            new_lr = max(old_lr * self.scheduler.factor, self.scheduler.min_lrs[i])
+            if old_lr - new_lr > self.scheduler.eps:
+<<<<<<< HEAD
+                param_group["lr"] = new_lr
+=======
+                param_group['lr'] = new_lr
+>>>>>>> b7dc9211c185a80dc75f0cd1db9bce07e6b648e0
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:1521]: you have to solve this merge issue

File: ignite/handlers/param_scheduler.py [hunk 2]
```diff
@@ -1393,6 +1393,133 @@ def simulate_values(cls, num_events: int, schedulers: List[_LRScheduler], **kwar
             return values
 
 
+class ReduceLROnPlateauScheduler(ParamScheduler):
+    """Reduce LR when a metric stops improving.
+    Wrapper of torch.optim.lr_scheduler.ReduceLROnPlateau .
+    Args:
+        optimizer (Optimizer): Wrapped optimizer.
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:1400]: If this docs are just copied from pytorch docs, let's remove it and provide the link to it.

Comment (sadra-barikbin) [reply ignite/handlers/param_scheduler.py:1400]: Hi, No. Parameters `metric_name`, `trainer`, `save_history` and `param_group_index` are not there. The example also is not from there.

Comment (vfdev-5) [reply ignite/handlers/param_scheduler.py:1400]: I see, then we can remove type hints from docsting as they will be fetched automatically from typehints from code

Comment (sadra-barikbin) [reply ignite/handlers/param_scheduler.py:1400]: Other parameters are copied from there. You mean to remove just those params?

File: ignite/handlers/param_scheduler.py [hunk 3]
```diff
@@ -1393,6 +1393,133 @@ def simulate_values(cls, num_events: int, schedulers: List[_LRScheduler], **kwar
             return values
 
 
+class ReduceLROnPlateauScheduler(ParamScheduler):
+    """Reduce LR when a metric stops improving.
+    Wrapper of torch.optim.lr_scheduler.ReduceLROnPlateau .
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:1398]: Let's add a link to pytorch docs for this class.

File: ignite/handlers/param_scheduler.py [hunk 4]
```diff
@@ -1393,6 +1393,133 @@ def simulate_values(cls, num_events: int, schedulers: List[_LRScheduler], **kwar
...
+    Args:
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:1399]:
```suggestion
    Wrapper of torch.optim.lr_scheduler.ReduceLROnPlateau .

    Args:
```

File: ignite/handlers/param_scheduler.py (hunk 2 above)
Comment (vfdev-5) [reply ignite/handlers/param_scheduler.py]: No, i was talking only about type hints

File: ignite/handlers/param_scheduler.py [hunk 5]
```diff
@@ -1393,6 +1393,133 @@ def simulate_values(cls, num_events: int, schedulers: List[_LRScheduler], **kwar
...
+        patience: int = 10,
+        threshold: float = 1e-4,
+        threshold_mode: str = "rel",
+        cooldown: int = 0,
+        min_lr: Union[float, List[float]] = 0.0,
+        eps: float = 1e-8,
+        trainer: Engine = None,
+        save_history: bool = False,
+        param_group_index: int = None,
+    ):
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:1473]: I think we can make constructor more simple. Keep only optimizer, trainer, save_history, param_group_index. Remove others and add **scheduler_kwargs which is passed to `ReduceLROnPlateau`. In docs we explain that **scheduler_kwargs are args/kwargs for `ReduceLROnPlateau`

File: ignite/handlers/param_scheduler.py [hunk 6]
```diff
@@ -1393,6 +1393,133 @@ def simulate_values(cls, num_events: int, schedulers: List[_LRScheduler], **kwar
...
+        if param_group_index:
+            if not isinstance(min_lr, float):
+                raise TypeError(f"When param_group_index is given, min_lr should be a float, but given {type(min_lr)}")
+        else:
+            _min_lr = min_lr
+            min_lr = [0] * len(optimizer.param_groups)
+            min_lr[param_group_index] = _min_lr
+
+        self.scheduler = ReduceLROnPlateau(
+            optimizer,
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:1490]: I'm not quite sure what happens if you pass optimizer here. How do you update only specified group `param_group_index` ?

Comment (sadra-barikbin) [reply ignite/handlers/param_scheduler.py:1490]: By replacing its `_reduce_lr` method which iterates over optimizer's param_groups and updates their LRs by the one of our own.

Comment (sadra-barikbin) [reply ignite/handlers/param_scheduler.py:1490]: Is it OK?

**Comment by sadra-barikbin** [comment]
Hi,

@vfdev-5 Anything else should I do?

File: ignite/contrib/handlers/__init__.py [hunk 7]
```diff
@@ -17,6 +17,7 @@
     LRScheduler,
     ParamGroupScheduler,
     PiecewiseLinear,
+    ReduceLROnPlateauScheduler,
```
Comment (vfdev-5) [inline ignite/contrib/handlers/__init__.py:20]: let's not include it to contrib module as contrib param scheduler handlers are deprecated

File: ignite/contrib/handlers/param_scheduler.py [hunk 8]
```diff
@@ -21,6 +21,7 @@
     ParamGroupScheduler,
     ParamScheduler,
     PiecewiseLinear,
+    ReduceLROnPlateauScheduler,
```
Comment (vfdev-5) [inline ignite/contrib/handlers/param_scheduler.py:24]: Same here. Let's clean up here.

File: ignite/handlers/param_scheduler.py [hunk 9]
```diff
@@ -1393,6 +1393,99 @@ def simulate_values(cls, num_events: int, schedulers: List[_LRScheduler], **kwar
             return values
 
 
+class ReduceLROnPlateauScheduler(ParamScheduler):
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:1396]: We have to include this class to the docs: https://github.com/pytorch/ignite/blob/master/docs/source/handlers.rst#parameter-scheduler

File: ignite/handlers/param_scheduler.py [hunk 10]
```diff
@@ -1393,6 +1393,99 @@ def simulate_values(cls, num_events: int, schedulers: List[_LRScheduler], **kwar
...
+    """Reduce LR when a metric stops improving.
+    Wrapper of torch.optim.lr_scheduler.ReduceLROnPlateau
+    <https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ReduceLROnPlateau.html>.
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:1399]: This does not look like a link, please check how it is rendered.

File: ignite/handlers/param_scheduler.py [hunk 11]
```diff
@@ -1393,6 +1393,99 @@ def simulate_values(cls, num_events: int, schedulers: List[_LRScheduler], **kwar
...
+            Must be attached the to same engine.
+        trainer: Trainer engine to log LR history in its
+            `state.output.param_history`. Is used if `save_history`
+            is true. Default: None.
+        save_history: Whether to save history or not. If true,
+            history will be logged in `trainer`'s `state.output.param_history`.
+            Default: False.
+        param_group_index: `optimizer`'s parameters group
+            to use.  Default: None. Use all `optimizer`'s paramater groups.
+        **scheduler_kwargs: Keyword arguments to be passed to the wrapped
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:1413]: Just a note: torch scheduler has `verbose` arg, but in our case, it seems like it wont be taken into account...
I'm not sure if it is good to use `print` as torch is doing now...

**Comment by vfdev-5** [comment]
Hi @sadra-barikbin thanks for updates, I left few other comments. I checked how it works and now i understand that you are patching an original private method : `self.scheduler._reduce_lr = self._reduce_lr` to make things to work. Let's say it is OK.

File: ignite/handlers/param_scheduler.py (hunk 11 above)
Comment (sadra-barikbin) [reply ignite/handlers/param_scheduler.py:1413]: I hard-coded it to `False` internally. We have `save_param_history` ourselves.

Comment (vfdev-5) [reply ignite/handlers/param_scheduler.py:1413]: Where it is done exactly ?

Comment (sadra-barikbin) [reply ignite/handlers/param_scheduler.py:1413]: Now I committed.

Comment (vfdev-5) [reply ignite/handlers/param_scheduler.py:1426]: Let's also mention that in the docs: verbose is set to False and user can use `save_history`.

File: ignite/handlers/param_scheduler.py [hunk 12]
```diff
@@ -1406,6 +1406,141 @@ def simulate_values(cls, num_events: int, schedulers: List[_LRScheduler], **kwar
...
+            if not isinstance(min_lr, float):
+                raise TypeError(f"When param_group_index is given, min_lr should be a float, but given {type(min_lr)}")
+            _min_lr = min_lr
+            min_lr = [0] * len(optimizer.param_groups)
+            min_lr[param_group_index] = _min_lr
+        else:
+            min_lr = 0
+        _scheduler_kwargs = scheduler_kwargs.copy()
+        _scheduler_kwargs["min_lr"] = min_lr
+        _scheduler_kwargs["verbose"] = False
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:1475]: @sadra-barikbin let's warn the user that we put verbose to False in case if `_scheduler_kwargs["verbose"] == True`, otherwise silent patching like that is not a good user experience. In the warning you can mention that user can use `save_history`

**Review by vfdev-5 — COMMENTED** [review]
@sadra-barikbin thanks for the updates and sorry for delay !
Few nits about verbose and we can merge this PR.

File: ignite/handlers/param_scheduler.py [hunk 13]
```diff
@@ -1406,6 +1406,141 @@ def simulate_values(cls, num_events: int, schedulers: List[_LRScheduler], **kwar
...
 
+class ReduceLROnPlateauScheduler(ParamScheduler):
+    """Reduce LR when a metric stops improving.
+    Wrapper of `torch.optim.lr_scheduler.ReduceLROnPlateau
+    <https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ReduceLROnPlateau.html>`_.
+
+    Args:
+        optimizer: Wrapped optimizer.
+        metric_name: metric whose improvement is monitored.
+            Must be attached the to same engine.
```
Comment (sdesrozis) [inline ignite/handlers/param_scheduler.py:1417]:
```suggestion
            Must be attached to the same engine.
```

File: ignite/handlers/param_scheduler.py [hunk 14]
```diff
@@ -1406,6 +1406,141 @@ def simulate_values(cls, num_events: int, schedulers: List[_LRScheduler], **kwar
...
+            history will be logged in `trainer`'s `state.output.param_history`.
+            Default: False.
+        param_group_index: `optimizer`'s parameters group
+            to use.  Default: None. Use all `optimizer`'s paramater groups.
+        **scheduler_kwargs: Keyword arguments to be passed to the wrapped
+            `ReduceLROnPlateau`.
+
+    Examples:
+
+        .. code-block:: python
```
Comment (sdesrozis) [inline ignite/handlers/param_scheduler.py:1431]: I think a doctest similar to others param scheduler should be added.

File: ignite/handlers/param_scheduler.py [hunk 15]
```diff
@@ -1406,6 +1406,141 @@ def simulate_values(cls, num_events: int, schedulers: List[_LRScheduler], **kwar
...
+                f"which itself has the metric {self.metric_name}."
+            )
+        self.scheduler.step(engine.state.metrics[self.metric_name])
+        super().__call__(self.trainer, name)
+
+    def get_param(self) -> Union[float, List[float]]:
+        lrs = [pg["lr"] for pg in self.optimizer_param_groups]
+        return lrs[0] if len(lrs) == 1 else lrs
+
+    def _reduce_lr(self, epoch: int) -> None:
```
[the remark this answers is not in the document]
Comment (vfdev-5) [reply ignite/handlers/param_scheduler.py:1495]: https://github.com/pytorch/ignite/pull/2449/files#diff-095d9f8e73275ad273d1e264433f83a92c6d8659a93543fc5355da7744aa28daR1478

There is a patching of original torch scheduler `self.scheduler._reduce_lr`. Without that it scheduler will update the all groups in the optimizer etc.

[the remark this answers is not in the document]
Comment (sdesrozis) [reply ignite/handlers/param_scheduler.py:1495]: Oh my bad, I hadn't seen the difference `self.optimizer_param_groups` vs `self.optimizer.param_groups`. I understand now why `min_lr` must be a float if `param_groups` is defined.

I wonder if it would have been better / safer / robust to wrap the optimizer instead of patching the scheduler.

[the remark this answers is not in the document]
Comment (vfdev-5) [reply ignite/handlers/param_scheduler.py:1495]: I propose to keep it like that and update later if there is a better idea.

**Comment by vfdev-5** [comment]
@sadra-barikbin can you resolve conflicts with master and address remaining comments. We can try to merge it then

**Review by vfdev-5 — APPROVED** [review]
Thanks for the update @sadra-barikbin !
Good to the PR.
